ticker/src/main/java/com/robinhood/ticker/TickerCharacterList.java

```diff
@@ -0,0 +1,114 @@
+ * it will know that it has to go from 'd' to 'c' to 'b', and these are the characters
+ * that show up during the animation scroll.
+ *
+ * @author Jin Cao, Robinhood
+ */
+public class TickerCharacterList {
+    private final int numOriginalCharacters;
+    // The saved character list will always be of the format: EMPTY, list, list
+    private final char[] characterList;
+    // A minor optimization so that we can cache the indices of each character.
+    private final Map<Character, Integer> characterIndicesMap;
+
+    public TickerCharacterList(String characterList) {
+        if (characterList.contains(Character.toString(TickerUtils.EMPTY_CHAR))) {
+            throw new IllegalArgumentException(
+                    "You don't need to include TickerUtils.EMPTY_CHAR in the character list.");
+        }
+
+        final char[] charsArray = characterList.toCharArray();
+        final int length = charsArray.length;
+        this.numOriginalCharacters = length;
+
+        characterIndicesMap = new HashMap<>(length);
+        for (int i = 0; i < length; i++) {
+            characterIndicesMap.put(charsArray[i], i);
+        }
+
+        this.characterList = new char[length * 2 + 1];
+        this.characterList[0] = TickerUtils.EMPTY_CHAR;
+        for (int i = 0; i < length; i++) {
+            this.characterList[1 + i] = charsArray[i];
+            this.characterList[1 + length + i] = charsArray[i];
+        }
+    }
+
+    /**
+     * @param start the character that we want to animate from
+     * @param end the character that we want to animate to
+     * @return a valid {@link AnimationCharacterIndices}, or null if the inputs are not supported.
+     */
+    AnimationCharacterIndices getCharacterIndices(char start, char end) {
+        int startIndex = getIndexOfChar(start);
+        int endIndex = getIndexOfChar(end);
+        if (startIndex < 0 || endIndex < 0) {
+            return null;
+        }
+
+        if (start != TickerUtils.EMPTY_CHAR && end != TickerUtils.EMPTY_CHAR &&
+                endIndex < startIndex) {
+            final int nonWrapDistance = startIndex - endIndex;
+            final int wrapDistance = numOriginalCharacters - startIndex + endIndex;
+            if (wrapDistance < nonWrapDistance) {
+                endIndex = numOriginalCharacters + endIndex;
+            }
+        }
+        return new AnimationCharacterIndices(startIndex, endIndex);
+    }
+
+    Set<Character> getSupportedCharacters() {
+       return characterIndicesMap.keySet();
+    }
+
+    char[] getCharacterList() {
+        return characterList;
+    }
+
+    private int getIndexOfChar(char c) {
+        if (c == TickerUtils.EMPTY_CHAR) {
+            return 0;
+        } else if (characterIndicesMap.containsKey(c)) {
+            return characterIndicesMap.get(c) + 1;
+        } else {
+            return -1;
+        }
+    }
+
+    public class AnimationCharacterIndices {
```

Comment:
this could just be a Pair<Integer, Integer>

Reply:
this is assuming we could also keep this class internal, which would be great

Reply:
i exposed this class to allow for overriding for custom behavior, but you are right using a pair is easier until other needs arise.

Reply:
actually im reverting back to use a custom class because `Pair` is an android class and it breaks my unit tests